src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs

```diff
@@ -3540,7 +3549,8 @@ private void AddSynthesizedConstructorsIfNecessary(ArrayBuilder<Symbol> members,
                     {
                         case MethodKind.Constructor:
                             // Ignore the record copy constructor
-                            if (!IsRecord || !SynthesizedRecordCopyCtor.HasCopyConstructorSignature(method))
+                            if (!IsRecord ||
+                                !(SynthesizedRecordCopyCtor.HasCopyConstructorSignature(method) && method is not SynthesizedRecordConstructor))
```

Comment:
Somehow, my brain just can't work out the exact purpose or meaning of this expression. When I apply De Morgan's to it (which I think makes it a little easier to understand) I get:

```cs
!IsRecord
|| !SynthesizedRecordCopyCtor.HasCopyConstructorSignature(method)
|| method is SynthesizedRecordConstructor
```

So we use a constructor as evidence to skip the implicit parameterless constructor if:
1. it's not contained in a record
2. it doesn't have the copy constructor signature, or
3. it is a synthesized record constructor.

I wasn't sure why (3) is necessary. If the record primary constructor gets skipped because it matches the copy constructor signature, isn't that fine, since it's an error scenario?

Reply:
> If the record primary constructor gets skipped because it matches the copy constructor signature, isn't that fine, since it's an error scenario?

No, it is not fine. We know it is not a copy constructor. We know it is a constructor explicitly declared in code, therefore, its presence should suppress declaration of a parameter-less constructor.

---
In reply to: [544673679](https://github.com/dotnet/roslyn/pull/49975#discussion_r544673679) [](ancestors = 544673679)